stor/dx.py

```diff
@@ -144,41 +138,273 @@ def download_objects(self):  # may not need it
+        dest_exists = dest.isdir()
+        if dest_exists:
+            new_dest = dest / self.name
+            if new_dest.isfile():
+                raise DuplicateError('Dest path ({}) already exists. Do not duplicate!'
+                                     .format(new_dest))
+            folder_dest = dest
+        else:
+            if dest.isfile():
+                raise DuplicateError('Dest path ({}) already exists. Do not duplicate!'
+                                     .format(dest))
+            folder_dest = dest.parent
+            folder_dest.makedirs_p()
+
+        file_handler.move(folder_dest.resource or '/')
+
+        if not dest_exists and not dest.endswith('/'):
+            file_handler.rename(dest.name)
+
+    def copy(self, dest, move_within_project=False):
+        """Copies data object to destination path.
+
+        Args:
+            dest (path|str): The destination file or directory.
+            move_within_project (bool): If True, make new files in DX instead of cloning.
+                Only comes into effect when both source and destination are DXPaths
+
+        Raises:
+            DNAnexusError : When cloning within same project with copy_new=False
+            DuplicateError : When cloning data object or folder to already existing path
+        """
+        dest = Path(dest)
+        if utils.is_dx_path(dest):
+            if self.isfile():
+                if dest.canonical_project == self.canonical_project:
+                    if move_within_project:
+                        self._move(dest)
+                    else:
+                        raise DNAnexusError('Cannot copy within same project without flag set')
+                else:
+                    self._clone(dest)
+            else:
+                raise NotFoundError('No data object was found for the given path on DNAnexus')
+        else:
+            super(DXPath, self).copy(dest)
 
     def download_object(self, dest):
-        """Download a single path or object to file."""
-        raise NotImplementedError
+        """Download a single path or object to file.
+
+        Args:
+            dest (Path): The output file
+
+        Raises:
+            ValueError: When source path is not a file
+        """
+        if not self.isfile():
+            raise ValueError('can only call download_object on object path')
+        dxpy.download_dxfile(
+            dxid=self.canonical_resource,
+            filename=dest,
+            project=self.canonical_project
+        )
 
     def download(self, dest):
-        """Download a directory."""
-        raise NotImplementedError
+        """Download a directory.
 
-    def upload(self, source):
-        """Upload a list of files and directories to a directory."""
-        raise NotImplementedError
+        Args:
+            dest (Path): The output directory
+
+        Raises:
+            ValueError: When source or dest path is not a directory
+        """
+        if not self.isdir():
```

Comment:
this check seems unnecessary. 

Also I think this needs to check that the dest is a filesystem path.

Reply:
Since this is always called for non-obs system (by utils.copy/copytree), the check would be redundant IMO. And the `isdir()` check will be unnecessary after I implement error parsing I think, because otherwise `dxpy.download_folder()` would throw a dxpy error, since it cannot find the folder

Reply:
seems like you need to update this according to your comment above.

Reply:
still seems like you don't need this check

Reply:
Fixed